crates/wascc-provider/src/lib.rs

```diff
@@ -437,21 +491,66 @@ struct VolumeBinding {
 fn wascc_run_http(
     host: Arc<Mutex<WasccHost>>,
     data: Vec<u8>,
-    env: EnvVars,
+    _env: EnvVars,
     volumes: Vec<VolumeBinding>,
     log_path: &Path,
     status_recv: Receiver<ContainerStatus>,
+    port_assigned: i32,
 ) -> anyhow::Result<ContainerHandle<ActorHandle, LogHandleFactory>> {
     let mut caps: Vec<Capability> = Vec::new();
 
+    let mut env_map: HashMap<String, String> = HashMap::new();
+    env_map.insert("PORT".to_string(), port_assigned.to_string());
     caps.push(Capability {
         name: HTTP_CAPABILITY,
         binding: None,
-        env,
+        env: env_map, //replaced _env
     });
     wascc_run(host, data, &mut caps, volumes, log_path, status_recv)
 }
 
+#[derive(Debug)]
+struct PortAllocationError {}
+
+impl PortAllocationError {
+    fn new() -> PortAllocationError {
+        PortAllocationError {}
+    }
+}
+impl fmt::Display for PortAllocationError {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "all ports are currently in use")
+    }
+}
+impl Error for PortAllocationError {
```

Comment:
Nit: We could probably just use the `anyhow` crate derives instead of manually implementing `Error`, but not going to block the PR on it